memory_allocator.c

```diff
@@ -0,0 +1,258 @@
+#include <common.h>
+#include <stdint.h>
+#include <stddef.h>
+#include "queue.h"
+#include "libkern.h"
+#include "include/memory_allocator.h"
+
+/*
+  TODO:
+  - use the mp_next field of malloc_pool
+*/
+
+#define MAGIC 0xC0DECAFE
+#define ALIGNMENT 8
+
+TAILQ_HEAD(mb_list, mem_block);
+TAILQ_HEAD(mp_arena, mem_arena);
+
+static inline int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
+typedef struct mem_block {
+  uint32_t mb_magic; /* if overwritten report a memory corruption error */
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
+  TAILQ_ENTRY(mem_block) mb_list;
```

Comment:
For debugging purposes you can scan all blocks in arena without using pointers anyway - just start at the beginning and move forward by `abs(blk->mb_size)`.

You're right union won't save any space, but the smallest usable block would be 16 bytes with header then (for up to 8 byte allocations, of course). I reckon it's worth considering since the implementation of that is trivial.